### PR TITLE
Add DICTIONARY

### PR DESCRIPTION
**Pull request by phoe**

### CODE REVIEW

File: dictionary.lisp [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+          (format T "~{  <li>~A</li>~%~}" glossaries)
+          (format T "</ol>~%")
+          (when synonyms
+            (format T "<p>Synonyms: ~{~{~%  <a href=#~A>~A</a>~}~^,~}~%</p>~%"
+                    (mapcar #'anchorize synonyms)))
+          (when antonyms
+            (format T "<p>Antonyms: ~{~{~%  <a href=#~A>~A</a>~}~^,~}~%</p>~%"
+                    (mapcar #'anchorize antonyms))))))))
+
+(define-widget browser (qtextbrowser)
```
Comment (Shinmera) [inline dictionary.lisp]: This name is too general, please change it to something that conveys this browser is specific to dictionaries.

File: dictionary.lisp [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+             ((old "QWidget*") (new "QWidget*"))
+  (declare (connected qt:*qapplication* (focus-changed "QWidget*" "QWidget*")))
+  (%set-focus-from-browser dictionary new))
+
+(defmethod set-focus-from-browser ((dictionary dictionary) old (new qobject))
+  (%set-focus-from-browser dictionary new))
+
+(defmethod set-focus-from-browser ((dictionary dictionary) old new))
+
+(defun %set-focus-from-browser (dictionary new)
```
Comment (Shinmera) [inline dictionary.lisp]: What's the point of this extra function? Why not just use the method?

File: dictionary.lisp [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+(in-readtable :qtools)
+
+(define-widget dictionary (qwidget)
+  ((empty-browser-text :accessor empty-browser-text
+                       :initarg :empty-browser-text)
+   (not-found-text :accessor not-found-text
+                   :initarg :not-found-text))
+  (:default-initargs
+    :empty-browser-text
+    "<i><p align=center>Type your query below and hit Search.</p></i>"
```
Comment (Shinmera) [inline dictionary.lisp]: This is not valid HTML. `p` is a block tag, and `i` is an inline tag, so the nesting should be switched.

File: dictionary.lisp [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    :not-found-text
+    "<i><p align=center>The entry for \"~A\" was not found.</p><i>"))
+
+(define-subwidget (dictionary layout) (q+:make-qgridlayout)
```
Comment (Shinmera) [inline dictionary.lisp]: If you `(q+:make-whateverlayout my-widget)` you don't need to `(setf (q+:layout ..) ..)` after. It's cleaner and shorter.

File: dictionary.lisp [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+  (setf (q+:layout dictionary) layout
+        (q+:contents-margins layout) (values 0 0 0 0)))
+
+(define-subwidget (dictionary browser)
+    (make-instance 'browser :dictionary dictionary)
+  (q+:add-widget layout browser 0 0 1 2)
```
Comment (Shinmera) [inline dictionary.lisp]: I much prefer defining the layout last and adding the widgets in the layout's initialiser, as the positioning is more layout-specific information.

**Review by Shinmera — CHANGES_REQUESTED** [review]

File: dictionary.lisp (hunk 1 above)
Comment (phoe) [reply dictionary.lisp]: Fixed.

File: dictionary.lisp (hunk 2 above)
Comment (phoe) [reply dictionary.lisp]: Fixed.

File: dictionary.lisp (hunk 3 above)
Comment (phoe) [reply dictionary.lisp]: Fixed.

File: dictionary.lisp (hunk 4 above)
Comment (phoe) [reply dictionary.lisp]: Fixed - TIL. I was not aware of this.

Comment (Shinmera) [reply dictionary.lisp]: You don't need to comment on each thing as fixed. If you push the patches it'll notice that it's changed automatically.

File: dictionary.lisp (hunk 5 above)
Comment (phoe) [reply dictionary.lisp]: Fixed.

**Review by Shinmera — APPROVED** [review]

File: dictionary.lisp (hunk 4 above)
Comment (phoe) [reply dictionary.lisp]: Woooo. I keep on learning GitHub stuff every day.

**Pull request merged**